### PR TITLE
[dipu] Optimize `getAllocator` by adopting lookup table

### PR DESCRIPTION
**Pull request by lljbash**
Slight improvement on ops like `empty`.
ResNet-18 (batch_size=1, iter=1000) 19.7s->19.6s

### CODE REVIEW

**Review by yewentao256 — APPROVED** [review]

**Review by fandaoyi — APPROVED** [review]

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.cpp [hunk 1]
```diff
@@ -98,6 +104,26 @@ c10::Allocator* getAllocator(const c10::Device& device) {
...
+c10::Allocator* getAllocator(const c10::Device& device) {
+  // allocator_lookup_table[device_type][device_index]
+  static std::array<
+      std::array<c10::Allocator*, allocator_details::kMaxDeviceNumber>,
+      static_cast<std::size_t>(c10::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)>
+      allocator_lookup_table{};
+  auto device_type = static_cast<int>(device.type());
+  int device_index =
+      device.type() == dipu::DIPU_DEVICE_TYPE
+          ? device.has_index() ? device.index() : devproxy::current_device()
```
[the remark this answers is not in the document]
Comment (lljbash) [reply dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.cpp:118]: 这 formatter 给排的 - -

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.h [hunk 2]
```diff
@@ -176,6 +176,8 @@ c10::Allocator* get_allocator_impl(c10::Allocator* raw_allocator) {
   return &cache_allocator;
 }
 
+constexpr int kMaxDeviceNumber = 16;
```
[the remark this answers is not in the document]
Comment (lljbash) [reply dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.h:174]: 确实

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lljbash) [reply dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUCachingAllocator.cpp:118]: 我直接改掉了

**Comment by lljbash** [comment]
@fandaoyi @wiryls 看下 [ae14e12](https://github.com/DeepLink-org/deeplink.framework/pull/532/commits/ae14e126a19e576e284a051cbea9cb90643705c2)

**Review by fandaoyi — APPROVED** [review]

**Pull request merged**